### ISSUE TITLE
folderMemoryCard: Should not call 8 times per frame just to increase a flush counter

### ISSUE THREAD

**Issue by avih** [PCSX2/pcsx2#838]
@AdmiralCurtiss I want to remove this line:
https://github.com/PCSX2/pcsx2/blob/f5a38ab/pcsx2/Counters.cpp#L437

It calls 8 times per frame into the plugin to increase a flush counter even if no folder memory cards are in use.

I want to remove this call completely.

Please explain to me what are the requirements here and let's find a solution without this call.

**Comment by avih** [comment]
@AdmiralCurtiss I've pushed ddb8702 to call once per second instead of 60 and also hacked around a hang issue at ef9016e .

You still need to remove this call completely, and please make sure that the other patch doesn't break the state of the MCDF.

Also, not sure if it's an issue or not, but I noticed that `PCSX2_UI.ini` has `SlotN_Type=1` for both normal memory card files and for memory cards folders. Is this how it's supposed to be? It seems to me rather redundant if it always ends up with the same value...

**Comment by AdmiralCurtiss** [comment]
The purpose of the flush counter is to let the memory card class know when the internal data is in a consistent state, so it can be written out to the file system. Consistency is assumed when the card has not been accessed for a certain amount of time. Since it's entirely possible for a game to write file data to an arbitrary offset of the memory card before it writes metadata that identifies what file the arbitrary data belongs to, I cannot write the data directly and must wait for a consistent memory card state. The call did not seem to impact performance, so I left it as-is, but if you can think of a better way to inform the card that it is consistent, feel free to change it.

I'll check out the empty file problem.

The ini file is wrong because the type is not used anymore, was removed with https://github.com/PCSX2/pcsx2/pull/687. Presumably you're seeing an entry that was added in one of the versions between #359 and #687.

**Comment by avih** [comment]
> I'll check out the empty file problem.

Thanks.

> Presumably you're seeing an entry that was added in one of the versions between #359 and #687.

Right. After I removed them, they didn't return. Thanks.

> The purpose of the flush counter is ..

Right, I realized what the purpose is after I followed the code and ended up with ddb8702 which at least calls it only once per sec.

> Consistency is assumed when the card has not been accessed for a certain amount of time

While this might hold true most of the time, it's not a good assumption IMO unless you treat it only as a hint and you're able to validate it somehow and abort if you detect that it turned out incorrect.

>  I cannot write the data to the host file system directly and must wait for a consistent memory card state before doing so

Why is that? Are there some states which cannot or should not be written to the host file system? I.e. real invalid states of the memory card? What happens if you assume the state is consistent but in practice it isn't, and you try to flush?

(these are not necessarily guiding questions. I ask because I don't know)

**Comment by AdmiralCurtiss** [comment]
Inconsistent data can't be (correctly) written to the host file system. The problem becomes clear with a simplified example.

PS2 Memory Cards store data in clusters, typically of 1 KB each. They use a file allocation table to remember which cluster belongs to which data, which includes both files and the file system itself. Write operations to the memory card are only atomic for block-sized areas, typically 16KB.

Let's assume we have a memory card consisting of two blocks. The first block holds our FAT and file system data, the second block is unused.

A game now wants to add a new file to the memory card. To do that, it has to:
* Store the actual file data.
* Update the file system data so that a reference to the just created data, including metadata like filename, filesize, and creation date, exists.
* Update the FAT so that it can later when reading the file back know which clusters contain file data in which order.

The game (or PS2 BIOS, I suppose) decides to store the file data in the second cluster. Since cluster writes are atomic, this shortly leaves the memory card in a state where file data has already been written (in the second cluster), but the file system and FAT (in the first cluster) has not yet been updated to reference that data. If we were to try and write this data to our host file system before that update happens, we would not be able to, because we don't have any information yet on which file this data belongs to.

This becomes even more complicated and confusing when data spans multiple clusters, or the FAT and file system aren't in the same cluster, or -- what happens in practice -- the file system data isn't in a nice block somewhere but spread randomly throughout the memory card. Ultimately, we just lack information about where written data belongs to if we try to write it to the host file system as it gets written by the game/PS2 to the memory card. We have to remember the data, wait for the game/PS2 to update the FAT and file system, then work backwards from that to figure out what files were actually written just now.


It's probably possible to validate some kind of consistency, but I don't know if it would work 100% of the time.

**Comment by avih** [comment]
What if we flush only after the FAT is updated? The file data should already be in place, right? I'm guessing that the FMC does know to read and interpret the FAT since it has to translate it into a file system on the host, so it shouldn't be hard to detect when the FAT was updated, I'd think.

Can the FAT itself be at an inconsistent state? Can this be detected?

**Comment by AdmiralCurtiss** [comment]
Hm, I wonder. Unfortunately the FAT of a typical memory card is bigger than a single block, but if we could reliably detect when the FAT is completely updated that would probably work. Assuming, of course, the PS2 always updates the FAT last -- I don't know if it does.

**Comment by avih** [comment]
While I'm not too familiar with file systems implementations, I'd like to think that implementations write the raw data it in a way that if aborted still stays consistent.

So Hopefully first write the file data while nothing points to it, then write the FAT entry pointing to the data but without updating the FAT itself, and then updating the FAT to include/point to the new FAT entry.

Assuming this is the order in which things happen, I'd think it shouldn't be impossible to detect this last stage of updating the FAT itself, and flush only and always when this happens (synchronously, i.e. without waiting a full second).

But this makes me think of another point. When you flush the data, you start with the FAT and work it from there (write all the dirs and files which the FAT points to), right?

So if there's some junk at the memory card which still doesn't have FAT entries pointing to it, then why would you consider it inconsistent or impossible to write? At the worst case, you'll flush the same content again (since the FAT hasn't been updated since the last flush), and if you're doing it smarter, you could read the file on the host before you try to write it, such that if you're going to write exactly the same data, just don't. Where's actually the issue here?

**Comment by AdmiralCurtiss** [comment]
Because that leads to reading and/or writing up to 8MB of data for every 16KB cluster the memory card writes, which is slow and inefficient.

Besides, there's still nothing that assures us that the data written at cluster x that is currently pointed to by file system entry y will remain part of that file when all is said and done. There's technically nothing stopping the PS2 from overwriting a file before the FAT/file system is updated to account for that overwriting, which in the extreme case leads to a game deciding to delete a file and create a different one in the same data cluster before updating the file name/path. I believe MGS3 specifically does that, in fact.

**Comment by avih** [comment]
> Because that leads to reading and/or writing up to 8MB of data for every 16KB cluster the memory card writes, which is slow and inefficient.

Slow and inefficient are very relative terms, and the goal here is to do the work, possibly slightly inefficiently, when required, while being able to avoid a permanent cost (and complexity) when it isn't required. Of course, these need to be balanced, and sometimes doing scheduled maintenance is fine, but I don't think this is the case here. If you claim slow and inefficient, some numbers to back your point would be useful here.

The OS keeps cache of recently accessed files, so reading 8M (if required) should be pretty fast if you've read or written this data recently. And besides, flushing only happens after writing to the card, which is very infrequent at the PS2.

> Besides, there's still nothing that assures us that the data written at cluster x that is currently pointed to by file system entry y will remain part of that file when all is said and done. There's technically nothing stopping the PS2 from overwriting a file before the FAT/file system is updated to account for that overwriting, which in the extreme case leads to a game deciding to delete a file and create a different one in the same data cluster before updating the file name/path. I believe MGS3 specifically does that, in fact.

Even if this happen, in what way does it break anything? So you're following the FAT and write a file with some clusters which were modified, and then on a later write this file is deleted or maybe remains the same after the FAT is updated. Why is this bad/invalid/inconsistent in any way?

**Comment by avih** [comment]
Also, please don't send a PR before we agreed on how to solve this (just a precaution). Thanks.

**Comment by avih** [comment]
Well?

**Comment by AdmiralCurtiss** [comment]
I'll get to it when I get to it. Frankly I don't see the hurry in changing this when the current method works fine and doesn't seem to cause any noticeable performance issues, especially when there's still outstanding actual (if minor) issues with the implementation I could tackle instead (see #666).

My approach would be to first add some basic logging so we can information on what order data gets written in, and if there's any standard order or if any game can just do whatever. Then based on that information we can probably figure out a sensible moment to Flush, perhaps even only flushing those area(s) of the memory card where we know the FAT has been updated to avoid garbage data being written.

Actually, in regards to that, it is important to not write garbage to deleted files because they're not actually deleted, but instead renamed. This allows the user to easily restore saves that have been accidentally deleted, either due to user error or due to actual errors in the game or emulator. Dolphin has the same feature, and it has helped me before, so I figured it would be good to have it here too.

**Comment by avih** [comment]
There's no hurry, but I want to resolve this.

> My approach would be to first add some basic logging ...

That's an optimization already - to reduce flush frequency. Let's forget optimizations until we understand what is correct/possible and what isn't, and once we nailed that, then we can then think how to optimize it.

> This allows the user to easily restore saves that have been accidentally deleted, either due to user error or due to actual errors in the game or emulator.

This sounds like a feature which goes beyond the scope of this discussion - a feature of "don't delete files form the host file system even if they're deleted from the memory card". If you actually planned this feature and have code which deals with it, we can add it as a factor to this discussion. Is this a planned feature?

There are basically few questions we need answered here:
1. Are there really bad states which are bad to flush?
2. If there are bad states, how do we make sure we don't flush them?
3. After everything works correctly, if it's too slow, think how to improve speed.

We're still at 1. Let's not discuss 2/3 until we've resolved 1.

I still didn't understand these
- What defines "invalid/bad state"?
- What happens if you flush it at this state?

Since the current system cannot guarantee to flush only when the memory card is "in a valid state", then we need to understand what happens if it flushes a bad state.

I.e. explain clearly what would break - from a user perspective. Would data get lost? Would the memory card folder become unusable?

**Comment by AdmiralCurtiss** [comment]
It is not an optimization, it is a prerequisite to knowing when the card writes are done. You simply cannot read a file system -- especially one that is as basic and without error prevention features as the PS2 memory card one -- that is in the middle of being rewritten and expect to get a sensible result. Anything could happen (data not being written correctly/in the correct location of a file/in the correct file/being discarded completely, whatever; it should at most affect files that are currently loaded in the card though) as long as new data happens to be in a spot that is reachable from the old or halfway modified FAT/File System but unrelated to that metadata. Hell if you're unlucky you could even end up with an infinitely recursing file system structure you have no good way to parse.

If you really want a definition, the best I can do is that an invalid state is the absence of a valid state. A valid state is a state where all internal metadata information such as the FAT and File System points at the data they're meant to point at. Sorry if this is unsatisfactory.

Of course I try to [make sure writes are done and abort if I can prove they're not](https://github.com/PCSX2/pcsx2/blob/master/pcsx2/gui/MemoryCardFolder.cpp#L855), but due to the way PS2 memory cards work you cannot guarantee it, and even a big and complicated analysis of the FAT/FS/data could easily end up with false positives.

Look, my first idea was not "cache all data in memory and write it when the game is done" either, but writing data directly simply turned out to not be a viable result. It took several ugly hacks to even get the card formatting operation to work correctly, and after that actually writing file data was basically impossible, it was [just a damn mess](https://github.com/AdmiralCurtiss/pcsx2/blob/bc9a6fc3a4219a2965aeb34afb9f4faffa091e4e/pcsx2/gui/MemoryCardFile.cpp#L1268). Admittedly, I understood less of the PS2 memory card file system at the time, but I don't think I would fare much better with my current knowledge. I just have to wait for a valid state and work from there. The **entire point** of the emulated-time-since-last-access feature you're trying to remove here is making as sure as we can that we have a valid state where all writes are completed, and if we want to remove that, we just have to find another way to confirm it.

I feel like we're sort of talking past each other here but I'm not sure what else I can say here.

Finally, the renaming instead of deleting is not a theoretical feature, it is implemented. It was implemented long before the original merge.

**Comment by avih** [comment]
> Consistency is assumed when the card has not been accessed for a certain amount of time

and

> I try to make sure writes are done and abort if I can prove they're not, but due to the way PS2 memory cards work you cannot guarantee it

and

> The entire point of the emulated-time-since-last-access feature you're trying to remove here is making as sure as we can that we have a valid state

You obviously went to great length to reduce the chance of flushing an invalid state, but the fact is you cannot guarantee it, so what happens if this assumption ends up not true? This is a real question which we should be able to answer. E.g. can it end up formatting the host hard drive? would it destroy the folder memory card forever? etc.

My main issue with this system is that it's a hack based on guestimation and some arbitrary delay to hopefully end up flushing in a valid state, rather than actually guaranteeing that the sate is valid when flushing, and we don't know what happens when this assumption ends up incorrect.

The fact that the implementation details of this delay are not fun is just an unfortunate side effect.

When we're writing to the host filesystem, we need to guarantee correctness first, and if we know we cannot guarantee it, then we must be able to tell what happens when things don't go to plan.

Once we know what could go wrong, then we can assess whether or not we can take the risk of flushing an invalid state.

But the situation right now is that you tried to explain how terrible would flushing an invalid state be, while also saying you cannot guarantee always flushing a valid state (the effort is good, but it's not guaranteed to work), and you still didn't say what would the side effect of such case be for the user. You basically hope for the best and ignore the case where "best" doesn't happen without knowing how it could end up.

Does this explain my issue with it?

**Comment by AdmiralCurtiss** [comment]
I can't answer the question of "what will happen" because it depends on the data. Most likely, it will discard file data (can't find file system entry for a given sector) or write it to a wrong but pre-existing file that is currently loaded in the memory card (finds the wrong file system entry since data is now in a place where another file's data was before). But I'm sure that there could be some bizarre combination of data and file system that has worse results, such as the file system being overwritten in a way that makes it look like all currently loaded files were deleted (one of the reasons for the rename failsafe), or the emulator locking up in an infinite loop trying to parse what just so happens to be a recursive file system. I doubt anything worse than that could happen, all writes should be limited to the folder memory card data. But again, this is theory. Since the game is free to write arbitrary data to the memory card I simply don't know what will happen exactly.

**Comment by avih** [comment]
> I simply don't know what will happen exactly.

> I'm sure that there could be some bizarre combination of data and file system that has worse results, such as the file system being overwritten in a way that makes it look like all currently loaded files were deleted (one of the reasons for the rename failsafe), or the emulator locking up in an infinite loop

We do agree that this is highly undesirable, right? So what do we do to make sure none of those ever happen?

**Comment by AdmiralCurtiss** [comment]
Sure, I absolutely agree that these things shouldn't happen. But I'm afraid I simply do not know how it would be possible to avoid them 100% of the time.

**Comment by AdmiralCurtiss** [comment]
And yeah, sure, the access counter is a check that is unlikely to catch all the cases ever. But that said I have yet to have an issue with it not correctly writing data using it.

**Comment by willkuer** [comment]
To my knowledge you can remove the memory card at any time during the game except for loading and writing scenes. And I think this happens as well on the fileMemoryCard. Maybe you could just eject the memory card before you flush it?

**Comment by avih** [comment]
The memory card code (and PCSX2 in general too IMO) doesn't know that it's a "loading or writing scene" or when one is over.

The problem is that right now there's no code which "knows when to flush it" other than a logic which says "hmm.. it's been a full second since the memory card was last accessed, let's assume it's OK to flush now".

Also, I don't see how ejecting the card before flush solves anything TBH.

**Comment by willkuer** [comment]
I think currently implementation in the file memory card works similarly as the folder memory card system. After some idle time the memcard is considered to be stable. In the file memory card system the memory card is ejected after that condition. If that is true and the bios wants to write something on the memory card but can not find it it either buffers the information or it disrcards it. Independent on this choice the file memory card system was working.

I guess that ejecting the memory card will prevent the bios to write wrong data but either discards it or buffers it until a memcard is inserted.

**Comment by avih** [comment]
Unless you can point to the code which does that, I don't think your description is correct. AFAIK the memory card is ejected automatically (and inserted) only after loading a state and only if the user has this feature enabled, in order for the _game_ itself to re-enumerate the content of the card - which might have changed between saving the state and loading it.

**Comment by avih** [comment]
> I guess that ejecting the memory card will prevent the bios to write wrong data but either discards it or buffers it until a memcard is inserted.

I still don't see how this would help the folder memory card code know whether it's OK or not OK to flush at any given point in time.

**Comment by AdmiralCurtiss** [comment]
Yeah, ejecting cards is not going to help here.

Now I am curious though, presumably memory card writes go through some BIOS function or at least some kind of SDK API, right? Can we perhaps hook into that somehow to find out when writes are done?

**Comment by willkuer** [comment]
I didn't want to provide a new check. The current implementation just waits sufficiently long. After that it is still questionable if saving is complete. Or if during writing of nonatomic blocks (as I've understood in principle all relevant writing accesses are non-atomic due to size limitations) there are updates which will invalidates the flush and possibly creates zombie data.

Ejecting the memory card will make all writing processes atomic.


I can not remember and can not check now but I think it is completely valid to remove the memory card during playing on the ps2. There needs to be some rescue algorithm if a game wants to save data during unusual times.

Again I am not providing a solution to check when the memory card is ready for flushing. I just propose that we don't need to know this if she just eject, flush, reinject possibly after some idle time as it is implemented now.

**Comment by willkuer** [comment]
But you can as well just ignore me for now. Obviously I am not sufficiently informed to help you much.

**Comment by gregory38** [comment]
is the issue still valid?

**Comment by avih** [comment]
Yes, still valid.

**Comment by AdmiralCurtiss** [comment]
Hey, I thought about this yesterday, and I have an idea. How about this: We only write to the file system on shutdown, and in the meantime, all intermediate writes go to a temporary file memory card constructed on bootup/index. If we find the temporary file on bootup, we know that we didn't properly shutdown last time and can try to recover data that hasn't yet been written to the file system.

Basically, what happens right now is:
- On boot, we index the file system and construct a memory card file system in memory.
- On write, we keep those writes in memory, wait a while, then flush that data out to the file system.
- On shutdown we don't do anything since writes are already done, or we flush if we haven't flushed yet since last write.

Instead, we could do:
- On boot, we index the file system and construct a full temporary memory card file on disk. This is gonna be slightly slower than the current construction but shouldn't be significant.
- On write, we just write to that temporary memory card file.
- On shutdown, we flush the card to the host file system and delete the temporary file.
- If we crash, we detect the temporary file on bootup and can ask the user if we should flush the data from that.

This eliminates the need of tracking time and flushing, while keeping all other properties of the current system (separate files in the host file system and automatic indexing based on game).

Thoughts?

**Comment by ISuckGood** [comment]
what happened here lol  ?

**Comment by lightningterror** [comment]
This should be resolved so closing.
